src/lib.rs

```diff
@@ -11,6 +11,678 @@
+        let props = UPClientMqtt::create_mqtt_properties_from_uattributes(attributes)?;
+        let data = message.payload.clone().unwrap(); // TODO: Handle optional case and is this the best way to send data?
+
+        let msg = mqtt::MessageBuilder::new()
+            .topic(topic)
+            .properties(props)
+            .payload(data)
+            .qos(1)
+            .finalize();
+
+        info!("Sending message: {:?}", msg);
+
+        // UserProperty is not sent with the message..
+
+        self.mqtt_client.publish(msg).await.map_err(|e| {
+            UStatus::fail_with_code(UCode::INTERNAL, format!("Unable to publish message: {e:?}"))
+        })?;
+
+        Ok(())
+    }
+
+    /// Add a UListener to an mqtt topic.
+    ///
+    /// # Arguments
+    /// * `topic` - Topic to add the listener to.
+    /// * `listener` - Listener to call when the topic recieves a message.
+    async fn add_listener(
+        &self,
+        topic: &str,
+        listener: Arc<dyn up_rust::UListener>,
+    ) -> Result<(), UStatus> {
+        let mut topic_listener_map = self.topic_listener_map.write().await;
+
+        if !topic_listener_map.contains_key(topic) {
+            // Subscribe to topic.
+            self.subscribe(topic).await?;
+        }
+
+        let listeners = topic_listener_map
+            .entry(topic.to_string())
+            .or_insert(HashSet::new());
+
+        // Add listener to hash set.
+        let comp_listener = ComparableListener::new(listener);
+        listeners.insert(comp_listener);
+
+        Ok(())
+    }
+
+    /// Remove a UListener from an mqtt topic.
+    ///
+    /// # Arguments
+    /// * `topic` - Topic to remove the listener from.
+    /// * `listener` - Listener to remove from the topic subscription list.
+    async fn remove_listener(
+        &self,
+        topic: &str,
+        listener: Arc<dyn up_rust::UListener>,
+    ) -> Result<(), UStatus> {
+        let mut topic_listener_map = self.topic_listener_map.write().await;
+
+        if topic_listener_map.contains_key(topic) {
+            topic_listener_map
+                .entry(topic.to_string())
+                .and_modify(|listeners| {
+                    // Remove listener from hash set.
+                    let comp_listener = ComparableListener::new(listener);
+                    listeners.remove(&comp_listener);
+                });
+
+            // Remove topic if no listeners are left.
+            if topic_listener_map.get(topic).unwrap().is_empty() {
+                topic_listener_map.remove(topic);
+
+                // Unsubscribe from topic.
+                self.unsubscribe(topic).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Helper function for subscribing the mqtt client to a topic.
+    ///
+    /// # Arguments
+    /// * `topic` - Topic to subscribe to.
+    async fn subscribe(&self, topic: &str) -> Result<(), UStatus> {
+        self.mqtt_client.subscribe(topic, 1).await.map_err(|e| {
```

Comment:
Again, just a thought, but putting QoS of 1 means yadda yadda might be good.

I guess they even have this available, which might be a little more descriptive too.
https://docs.rs/paho-mqtt/latest/paho_mqtt/types/constant.QOS_1.html

Reply:
Good callout on the constant